README.md
# snmp4j-srccode

Add some log purpose codes and test codes in snmp4j source code. this is integrated together with the source code and test code.
This will be helpful to diagnose some snmp4j problem.
Current snmp4j version is 2.5.6

Another snmp4j demo code repo is : https://github.com/gaoxingliang/snmp4jdemo


CHANGES.txt

CHANGE LOG OF SNMP4J
====================

[2017-03-06] Version 2.5.6:

* Fixed: Default security protocol ID was missing for PrivAES192With3DESKeyExtension.java
* Fixed: Possible NullPointerException in ArgumentParser.parse if not all format types have
  been provided.
* Improved: USM: security name in debug log is also printed in ASCII if printable (not only
  always as hex string as before).

[2017-02-06] Version 2.5.5:

* Fixed [SFJ-141]: OID.nextPeer() throws a StackOverflowException on OIDs with 127 sub-identifiers or more.
* Changed: Added SHA-224, and SHA-384 to the protocols added by SecurityProtocols.addDefaultProtocols().

[2017-01-02] Version 2.5.4:

* Fixed [SFJ-136]: SHA384 and SHA512 authentication protocols used block size of 64 instead 128
  as required by RFC 6234, section 6.4, page 14.

[2016-11-15] Version 2.5.3:

* Fixed [SFJ-133]: DefaultUdpTransportMapping getListenAddress() does not return the local IP address.
* Fixed [SFJ-132]: TableUtils is caught in a notify when agent does not return VBs in lexicographic order.
* Fixed [SFJ-134]: DefaultUdpTransportMapping listen() fails to run if socket timeout is non-zero.

[2016-09-10] Version 2.5.2:

* Fixed: Java compiler/source version reset to 1.6 (instead 1.8 in 2.5.1).

[2016-09-10] Version 2.5.1:

* Improved: DefaultPDUFactory default constructor now sets context engine ID and name with default values.
* Fixed [SFJ-130]: SNMP4J sends GETNEXT PDU with error-index == 1 when auto-converting GETBULK to GETNEXT.
* Fixed: Many JavaDoc syntax errors reported by Java 9 by default.

[2016-05-02] Version 2.5.0:

* Added [SFJ-120]: Maven Central deployment support (refactored pom.xml).
* Fixed [SFJ-118]: DefaultTcpTransportMapping: SocketEntry instances created for incoming TCP connections
  are not immediately removed when connection is closed remotely.
* Fixed [SFJ-119]: Snmp: NotificationListener cannot be registered with custom configured TransportMapping.
* Fixed [SFJ-121]: DefaultPDUFactory.createPDU does not set contextName and contextEngineID on ScopedPDU
  instances.
* Fixed [SFJ-124]: ArgumentParser cannot parse formats with default values for parameters correctly.
* Added: Added MaxAccess, SubIndexInfo, and SubIndexInfoImpl to better support MIB specification
  based SNMP operations.
* Improved: Added custom OID configuration support for non-standard security protocols in
  SecurityProtocols.addDefaultProtocols.
* Added PrivGeneric to reduce code size.
* Added SecurityPrivacy.isSupported() to be able to test security protocol support by JVM.
* Fixed return parameter type of ArgumentParser.parse(..).

[2016-02-20] Version 2.4.3:

* Fixed: Socket close after bind failure in DefaultTcpTransportMapping.
* Changed: snmp4jStatsReqTableWaitTime now provides the peer Address instead of the target information
  as index value to conform with other counters of the same table.
* Added: RequestStatistics interface to support optional collection of statistic values during
  message processing. The RequestStatistics will be returned by the Snmp class to CommandResponder listeners
  via the CommandResponderEvent object's getPduHandle() method if the MessageDispatcher used created
  a PduHandle that implements the RequestStatistics interface.
* Improved: Initial msgID generation according to recommendation of RFC 3412 §6.2.
* Added: Counter constants for message runtime measurement.

[2016-01-25] Version 2.4.2:

* Fixed: Removed System.out.println calls in TableUtils.
* Fixed [SFJ-117]: DefaultTcpTransportMapping does not accept incoming messages if their SNMP header
  (<= first 6 bytes) is fragmented.
* Added: Support for table based Counters.
* Added: Support for retry, timeout and waitTime counters in class Snmp based on the
  SNMP4J-STATISTICS-MIB using the CounterSupport events.

[2015-12-30] Version 2.4.1:

* Fixed: SHA-2 authentication protocols used wrong MAC and key lengths (were switched accidentally
  in version 2.4.0.
* Fixed: Compilation warning in TransportMappings due to generic refactorings in version 2.4.0.

[2015-12-21] Version 2.4.0:

* Added [SFJ-112]: Added SHA-2 (SHA-224,-256,-384,-512) support to USM as defined by RFC 7630.
* Fixed [SFJ-116]: Improve performance of TableUtils getTable methods for sparse tables.
* Fixed [SFJ-115]: USM: Allow  adjustments of the local engine time.
* Improved: TransportStateEvent now provides the messages that are being discarded due to an IO
  issue with a TCP connection.
* Improved: Time information is updated also from first response of an agent.
* Improved: Engine ID caching for target addresses. Now only confirmed class PDUs are cache for
  outgoing messages.
* Improved: Removed adding of engine IDs from traps to local MPv3 engine ID cache.
* Improved: Removed comments with obsolete code from the TLSTM.
* Improved: Small refactoring in Snmp class.
* Improved: Generics refactorings in MultiThreadedTrapReceiver and LogControl.
  TransportMapping address parameter type (now with wildcard).

[2015-07-18] Version 2.3.4:

* Fixed [SFJ-110]: Security fix: SNMP4J MPv3 adds engine IDs to the local cache even if they are from a
  non-authoritative source.
* Improved: The local engine ID cache now has an upper limit of 50.000 by default. Additional
  engine IDs for new addresses are rejected if this limit is reached. The limit can be changed and also
  disabled by SNMP4JSettings.setMaxEngineIdCacheSize.
* Added: MPv3.getEngineIdCacheSize() to be able to monitor the size of the MPv3 engine ID cache.
* Added: USM error codes for too long engine ID and too long security name.
* Fixed [SFJ-107]: Reusing a UDP socket after closing the DefaultUdpTransportMapping is not immediately
  possible if no socket timeout (=0) is used.
* Fixed [SFJ-103]: The DefaultUdpTransportMapping ListenThread does not exit if no socket timeout is set
  and a socket exception is caught.
* Improved: JavaDoc of Variable.clone() method.

[2014-12-19] Version 2.3.3:


src/main/java/ParseMessage.java
import org.snmp4j.MessageDispatcherImpl;
import org.snmp4j.SNMP4JSettings;
import org.snmp4j.TransportMapping;
import org.snmp4j.asn1.BERInputStream;
import org.snmp4j.log.LogFactory;
import org.snmp4j.mp.MPv3;
import org.snmp4j.security.*;
import org.snmp4j.smi.AbstractVariable;
import org.snmp4j.smi.OctetString;
import org.snmp4j.smi.UdpAddress;
import org.snmp4j.transport.DefaultUdpTransportMapping;

import java.net.InetAddress;
import java.nio.ByteBuffer;

/**
 * a debug class we could parse the raw packet message for snmp v3:
 *
 *    1) the input message bytes is in the output of the message:
 *          Option 1: With the following output when the snmp4j log is enabled:
 *              Received message from /10.10.10.10/161 with length 156: 30:81:99:.....XXXX
 *          Option 2: We can copy the hex stream from wireshark in:
 *                  Simple Network Management Protocol -> right click Copy -> ... as a Hex Stream
 *
 *
 *    2) the related username and password must be correct set if it's snmp v3
 *
 * //todo: add support for v2 message parsing
 */
public class ParseMessage {
    public static void main(String[] args) throws Exception {
        System.setProperty(LogFactory.SNMP4J_LOG_FACTORY_SYSTEM_PROPERTY, DebuggerLogFactory.class.getCanonicalName());

        /**
         * register a special integer 32 parser to avoid the length check
         */
        SNMP4JSettings.setExtensibilityEnabled(true);
        System.setProperty(AbstractVariable.SMISYNTAXES_PROPERTIES, "customizedsmisyntaxes.properties");

        // ~~~~

        MPv3 mpv3 = new MPv3();
        MessageDispatcherImpl messageDispatcher = new MessageDispatcherImpl();
        messageDispatcher.addMessageProcessingModel(mpv3);

        USM usm = new USM();

        UsmUser usmUser = new UsmUser(new OctetString("logicmonitor"), AuthSHA.ID, new OctetString("logicmonitor"), PrivAES128.ID, new OctetString("logicmonitor"));
        usm.addUser(usmUser);
        SecurityModels.getInstance().addSecurityModel(usm);

        // packet address
        UdpAddress addr =  new UdpAddress(InetAddress.getByName("127.0.0.1"), 161);

        // Option1: from snmp4j log:
        String message = "30:81:99:02:01:03:30:10:02:03:00:b3:94:02:03:00:ff:e3:04:01:03:02:01:03:04:43:30:41:04:11:80:00:1f:88:80:74:7a:d8:50:ad:42:f4:58:00:00:00:00:02:01:03:02:03:12:78:64:04:0c:6c:6f:67:69:63:6d:6f:6e:69:74:6f:72:04:0c:ba:93:35:3b:14:bb:ec:22:9a:b2:07:e9:04:08:ab:7b:11:4a:95:77:e5:54:04:3d:56:8d:55:58:6b:8a:09:17:88:e7:d7:18:89:42:4a:76:45:15:e2:ed:39:94:26:ce:8f:73:be:03:bf:e7:f3:6a:f6:61:96:9f:55:0a:52:0c:82:2f:d4:5b:30:49:aa:8c:b4:4b:a1:c2:d4:45:0c:8e:0b:b3:00:62:df";

        // Option2: from wireshark
        String message4Pcap = "3081990201033010020300b394020300ffe304010302010304433041041180001f8880747ad850ad42f458000000000201030203127864040c6c6f6769636d6f6e69746f72040cba93353b14bbec229ab207e90408ab7b114a9577e554043d568d55586b8a091788e7d71889424a764515e2ed399426ce8f73be03bfe7f36af661969f550a520c822fd45b3049aa8cb44ba1c2d4450c8e0bb30062df";
        StringBuffer stringBuffer = new StringBuffer();
        char [] chars = message4Pcap.toCharArray();
        for (int i = 0; i < chars.length / 2; i = i+2) {
            stringBuffer.append(chars[i]).append(chars[i+1]).append(":");
        }
        stringBuffer.deleteCharAt(stringBuffer.length() - 1);
        System.out.println("From wireshark:" + stringBuffer);


        byte [] bytes = OctetString.fromHexString(message, ':').getValue();
        ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
        BERInputStream in = new BERInputStream(byteBuffer);

        TransportMapping transportMapping = new DefaultUdpTransportMapping();
        messageDispatcher.processMessage(transportMapping, addr, in, null);

    }
}
